File: public/core/src/main/java/se/infomaker/iap/ui/util/UI.java
package se.infomaker.iap.ui.util;

import android.content.res.Resources;
import android.view.View;
import android.view.ViewGroup;

import java.util.HashMap;
import java.util.Map;

import timber.log.Timber;

public class UI {
    /**
     * Extract all views with id from view and put in map with id string as key
     * @param view to extract views from
     * @return map of views in view (including the view)
     */
    public static Map<String,View> extractViews(View view) {
        HashMap<String, View> viewMap = new HashMap<>();
        extractViews(viewMap, view);
        return viewMap;
    }

    private static void extractViews(Map<String, View> target, View view) {
        try {
            if (view.getId() != View.NO_ID) {
                String name = view.getResources().getResourceEntryName(view.getId());
                if (name != null) {
                    target.put(name, view);
                }
            }
        }
        catch (Resources.NotFoundException e) {
            Timber.e(e, "Could not find view");
        }
        if (view instanceof ViewGroup) {
            ViewGroup group = (ViewGroup) view;
            for (int i = 0; i < group.getChildCount(); i++) {
                extractViews(target, group.getChildAt(i));
            }
        }
    }
}

File: public/live-content/src/main/java/se/infomaker/iap/articleview/util/UI.kt
package se.infomaker.iap.articleview.util

import android.content.res.Resources
import android.view.View
import android.view.ViewGroup
import com.navigaglobal.mobile.livecontent.R
import timber.log.Timber
import java.util.ArrayList
import java.util.HashSet

object UI {
    fun px2dp(px: Float): Float {
        val metrics = Resources.getSystem().displayMetrics
        val dp = px / (metrics.densityDpi / 160f)
        return Math.round(dp).toFloat()
    }

    fun dp2px(dp: Float): Float {
        val metrics = Resources.getSystem().displayMetrics
        val px = dp * (metrics.densityDpi / 160f)
        return Math.round(px).toFloat()
    }

    fun View.getAllChildren(): List<View> {
        val objects = HashSet<View>()
        addAllChildrenUnique(objects)
        return ArrayList(objects)
    }

    private fun View.addAllChildrenUnique(set: MutableSet<View>) {
        if (this !is ViewGroup) {
            set.add(this)
            return
        }
        set.add(this)
        for (i in 0 until this.childCount) {
            val child = this.getChildAt(i)
            set.add(this)
            child.addAllChildrenUnique(set)
        }
    }

    fun View.mapSubViews() :Map<String, View> {
        val namedViews = mutableMapOf<String, View>()
        this.getAllChildren()
                .filter { it.id != -1 }
                .forEach {
                    try {
                        val name = it.resources.getResourceEntryName(it.id)
                        if (name != null) {
                            namedViews.put(name, it)
                        }
                    } catch (e: Resources.NotFoundException) {
                        Timber.w(e, "did not find resource")
                    }
                }
        this.setTag(R.id.viewMap, namedViews)
        return namedViews
    }
}